packages/graph/hash_graph/lib/graph/src/store/query/filter.rs

```diff
@@ -189,17 +225,115 @@ impl<'q, T: QueryRecord> TryFrom<Expression> for Option<FilterExpression<'q, T>>
 #[derive(Debug, PartialEq, Deserialize)]
 #[serde(untagged)]
 pub enum Parameter<'q> {
+    Boolean(bool),
     Number(f64),
     Text(Cow<'q, str>),
-    Boolean(bool),
+    #[serde(skip)]
+    Uuid(Uuid),
+    #[serde(skip)]
+    SignedInteger(i64),
+}
+
+impl Parameter<'_> {
+    fn to_owned(&self) -> Parameter<'static> {
+        match self {
+            Parameter::Boolean(bool) => Parameter::Boolean(*bool),
+            Parameter::Number(number) => Parameter::Number(*number),
+            Parameter::Text(text) => Parameter::Text(Cow::Owned(text.to_string())),
+            Parameter::Uuid(uuid) => Parameter::Uuid(*uuid),
+            Parameter::SignedInteger(integer) => Parameter::SignedInteger(*integer),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[must_use]
+pub struct ParameterConversionError {
+    actual: Parameter<'static>,
+    expected: ParameterType,
+}
+
+impl fmt::Display for ParameterConversionError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "could not convert `{}` to {}",
+            self.actual, self.expected
+        )
+    }
+}
+
+impl Context for ParameterConversionError {}
+
+impl Parameter<'_> {
+    #[expect(clippy::match_same_arms, reason = "Add validation of parameters")]
+    fn convert_to_parameter_type(
+        &mut self,
+        expected: ParameterType,
+    ) -> Result<(), Report<ParameterConversionError>> {
+        match (&mut *self, expected) {
+            (_, ParameterType::Any)
+            | (Parameter::Boolean(_), ParameterType::Boolean)
+            | (Parameter::Number(_), ParameterType::Number)
+            | (Parameter::Text(_), ParameterType::Text) => {
+                // no action needed, exact match
+            }
+            (Parameter::Text(_base_uri), ParameterType::BaseUri) => {
+                // TODO: validate base uri
+            }
+            (Parameter::Text(_versioned_uri), ParameterType::VersionedUri) => {
+                // TODO: validate versioned uri
+            }
+            (_, ParameterType::Timestamp) => {
+                // TODO: validate timestamps
+            }
+            (Parameter::Text(text), ParameterType::Uuid) => {
+                *self = Parameter::Uuid(Uuid::from_str(&*text).into_report().change_context_lazy(
+                    || ParameterConversionError {
+                        actual: self.to_owned(),
+                        expected: ParameterType::Uuid,
+                    },
+                )?);
+            }
+            (Parameter::Number(number), ParameterType::UnsignedInteger) => {
+                // Postgres cannot represent unsigned integer, so we use i64 instead
```

Comment:
Is there a way this can go wrong? Can people pass in negative numbers where we'll think it's an integer, or will we always think it's a `number`?

Reply:
Or better phrased, will the error below always be a mistake in the query, rather than a bug from us?

Reply:
It would be a bug in the query. `UnsignedInteger` is used for ontology versions, which we expect to be always positive.

Reply:
I'm assuming it would also be used for `minItems` and such?

Reply:
I haven't thought about that as we just use the JSON blob, but that would be used for that as well, yes 🙂